Xamarin.Forms.Core/Binding.cs

```diff
@@ -113,23 +116,80 @@ internal override void Apply(bool fromTarget)
 			_expression.Apply(fromTarget);
 		}
 
-		internal override void Apply(object context, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false)
+		internal override void Apply(object context, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false, bool fromAncestorChanged = false)
 		{
 			object src = _source;
 			var isApplied = IsApplied;
 
 			base.Apply(src ?? context, bindObj, targetProperty, fromBindingContextChanged: fromBindingContextChanged);
 
-			if (src != null && isApplied && fromBindingContextChanged)
+			if (isApplied && !ShouldReapply(fromBindingContextChanged, fromAncestorChanged))
 				return;
 
-			object bindingContext = src ?? Context ?? context;
+			if ( this.Source is RelativeBindingSource )
+				ResolveRelativeSource(bindObj, targetProperty);
+			else
+				CompleteApplyBinding(src ?? Context ?? context, bindObj, targetProperty);
+		}
+
+		private void CompleteApplyBinding(object bindingContext, BindableObject bindObj, BindableProperty targetProperty)
+		{
 			if (_expression == null && bindingContext != null)
 				_expression = new BindingExpression(this, SelfPath);
-
 			_expression.Apply(bindingContext, bindObj, targetProperty);
 		}
 
+		private async void ResolveRelativeSource(BindableObject bindObj, BindableProperty targetProperty)
+		{
+			RelativeBindingSource relSource = this.Source as RelativeBindingSource;
+			object bindingContext = null;
+			switch (relSource.Mode)
+			{
+				case RelativeBindingSourceMode.Self:
+					bindingContext = bindObj;
+					break;
+				case RelativeBindingSourceMode.TemplatedParent:
+					{
+						var view = bindObj as Element;
+						if (view == null)
+							throw new InvalidOperationException();
+						bindingContext = await TemplateUtilities.FindTemplatedParentAsync(view);
```

Comment:
can you explain why this is async ? and if it's right to have it async, you'll need `ConfigureAwait(false)`

Reply:
Yeah so this was how it was being done before, just inside `TemplateBinding`. `TemplateBinding.Apply` was an `async void`. The ultimate culprit is `TemplateUtilities.GetRealParentAsync` (which `FindAncestor` also needs) and that is using a `TaskCompletionSource` and an event hook in order to `await` the setting of the parent, presumably because the binding might be applied before the parent is set. 

I'm assuming there are other ways around this, but I was trying to change as little as necessary. I'll give it a shot if you think it's worth trying to eliminate, but it looks like template binding has worked this way for a few years now so I don't know if you want me to mess with it.